### ISSUE TITLE
[iOS only]Can't open sheet inside modal

### ISSUE THREAD

**Issue by ilovinht** [ammarahm-ed/react-native-actions-sheet#217]
It works on Android, but sometimes the sheet opens with extra space in the bottom.
And on ios, when I tap the button which opens the sheet, nothing happens and app freeze after closing the modal...
I'm new to RN. Is this a bug or my mistake...?

my modal.js
```
import React, {useState, useRef, useEffect} from 'react';
import {
  StyleSheet,
  SafeAreaView,
  Share,
  BackHandler,
  Platform,
  Dimensions,
} from 'react-native';
import Modal from 'react-native-modal';
import ProgressWebView from 'react-native-progress-webview';
import Header from './Header';
import ActionSheet, {
  ActionSheetRef,
  SheetManager,
  SheetProps,
  SheetProvider,
} from 'react-native-actions-sheet';
import ExtraDimensions from 'react-native-extra-dimensions-android';
import { actionSheetRef } from './actionSheet';
const H =Platform.select({
  android:ExtraDimensions.get('REAL_WINDOW_HEIGHT'),
 ios: Dimensions.get('screen').height,
  });
const WebViewModal = ({isVisible = false, handleModal, uri}) => {
  const webviewRef = useRef();
  const [canGoBack, setCanGoBack] = useState(false);
  const [currentUrl, setCurrentUrl] = useState();
  const handleBackButtonPress = () => {
    if (webviewRef.current) {
      webviewRef.current.goBack();
      return true;
    }
    return false;
  };
  useEffect(() => {
    BackHandler.addEventListener('hardwareBackPress', handleBackButtonPress);
    return () => {
      BackHandler.removeEventListener(
        'hardwareBackPress',
        handleBackButtonPress,
      );
    };
  }, []);
  const backButtonHandler = () => {
    if (webviewRef.current) {
      webviewRef.current.goBack();
    }
  };
  const onShare = async () => {
    /* actionSheetRef.current?.show();
    console.log(actionSheetRef.current?.isOpen()); */
    SheetManager.show('sheet', {
      payload: {url: currentUrl, ref: webviewRef},
    });
  };
  return (
    <Modal
      deviceHeight={H}
      backdropColor="transparent"
      isVisible={isVisible}
      style={styles.modal}
      animationIn="fadeIn"
      animationOut="fadeOut"
      onBackButtonPress={backButtonHandler}>
      <SafeAreaView style={{flex: 1, backgroundColor: '#FFFFFF'}}>
        <Header
          rightVisible={false}
          canClose
          onClose={handleModal}
          onBack={backButtonHandler}
          canGoBack={canGoBack}
          showShareActionSheet={onShare}
        />
        <ProgressWebView
          pullToRefreshEnabled
          source={{
            uri: uri,
          }}
          ref={webviewRef}
          onNavigationStateChange={navState => {
            setCanGoBack(navState.canGoBack);
            // setCanGoForward(navState.canGoForward);
            setCurrentUrl(navState.url);
          }}
          color={'#333333'}
          sharedCookiesEnabled={true}
          thirdPartyCookiesEnabled={true}
          useWebKit
        />
      </SafeAreaView>
    </Modal>
  );
};
const styles = StyleSheet.create({
  modal: {
    margin: 0,
  },
});
export default WebViewModal;
```

my sheet.tsx
```
import React, {useState, useRef, useEffect} from 'react';
import ActionSheet, {
  ActionSheetRef,
  SheetManager,
  SheetProps,
  registerSheet,
} from 'react-native-actions-sheet';
import {
  StyleSheet,
  SafeAreaView,
  Share,
  BackHandler,
  View,
  Text,
  Button,
  Linking,
  Pressable,
  TouchableOpacity,
} from 'react-native';
function Sheet(props: SheetProps) {
  const actionSheetRef = useRef<ActionSheetRef>(null);
  return (
    <ActionSheet id={props.sheetId} ref={actionSheetRef}>
      <View
        style={{zIndex:1000,
          justifyContent: 'center',
          alignItems: 'center',
          paddingVertical: 8,
        }}>
        <TouchableOpacity
          style={styles.button}
          onPress={async () => {
            await Linking.canOpenURL(props.payload?.url);
            Linking.openURL(props.payload?.url);
            SheetManager.hide('sheet');
          }}>
          <Text style={styles.text}>open in browser</Text>
        </TouchableOpacity>
        <View style={styles.separator} />
        <TouchableOpacity
          style={styles.button}
          onPress={() => {
            props.payload?.ref.current.reload();
            SheetManager.hide('sheet');
          }}>
          <Text style={styles.text}>reload</Text>
        </TouchableOpacity>
        <View style={styles.separator} />
        <TouchableOpacity
          style={{
            width: '100%',
            alignItems: 'center',
            paddingVertical: 18,
            backgroundColor: '#FFF',
            activeOpacity: 0.5,
          }}
          onPress={() => {
            SheetManager.hide('sheet');
          }}>
          <Text style={styles.text}>Cancel</Text>
        </TouchableOpacity>
      </View>
    </ActionSheet>
  );
}
registerSheet('sheet', Sheet);
const styles = StyleSheet.create({
  text: {
    fontSize: 16,
    fontFamily: 'NotoSerifJP-Regular',
    lineHeight:20,
    color:'#000',
  },
  button: {
    width: '100%',
    activeOpacity: 0.5,
    alignItems: 'center',
    backgroundColor: '#FFF',
    paddingVertical: 18,
  },
  separator: {
    width: '85%',
    borderBottomColor: '#CBCBCB',
    borderBottomWidth: StyleSheet.hairlineWidth,
  },
});
export {Sheet};
```

I've wrapped app with SheetProvider in app.js like:
```
<Provider store={store}>
      <SheetProvider>
        <NativeBaseProvider colorModeManager={colorModeManager}>
          <NavigationContainer ref={navigationRef}>
            <StatusBar backgroundColor="#fff" barStyle="dark-content" />
            <RootNavigator />
          </NavigationContainer>
        </NativeBaseProvider>
      </SheetProvider>
    </Provider>
```

**Comment by johnny-mcfadden-dailypay** [comment]
Just bumped into the same issue with more or less the same component hierarchy as above. Temporary work around seems to be not registering the sheet inside the modal and just rendering it as a normal component. 


/* IMPORTANT - For action sheets used inside Modals, registering sheets doesn't seem to work
  - It may be where we have our SheetProvider wrapping
  - Workaround is just to not register it and render the action sheet as a normal component
*/

**Comment by ammarahm-ed** [comment]
@johnny-mcfadden-dailypay @ilovinht use multiple SheetProviders. Register a new SheetProvider inside the opened Sheet component then register your sheet with your global Sheet Provider + Local. And then it should work as expected when Opening sheet with SheetManager in the given context. It's probably on the docs too.

**Comment by johnny-mcfadden-dailypay** [comment]
Ahh! I'll give that a go 🔥 Just to confirm:

- Put a new sheet provider inside the modal? 
- Import a new locally scoped sheets file into the modal that registers the sheets displayed inside the modal?

**Comment by johnny-mcfadden-dailypay** [comment]
@ammarahm-ed adding a SheetProvider as a child in the modal seemed to work a treat, thank you v much.

**Comment by johnny-mcfadden-dailypay** [comment]
@ammarahm-ed just one thing to note incase someone else bumps into it. I noticed the freeze behaviour still occurs in the following scenario:

- Modal has a sheet provider child wrapping all the modal children, where one of the children makes  a call to show the action sheet:

```
<Modal>
  <SheetProvider> 
   {children} // some child makes a call to open an action sheet
</SheetProvider>
</Modal>
```
- The action sheet is now open. In this scenario it's an action sheet that says: "Are you sure you want to exit this modal flow" with buttons saying 'no stay' and 'yes close'. 
- The 'yes close' button was just closing the modal itself.
- The modal close call was made, but the action sheet call wasn't made. As a result the modal slides down, the action sheet remains visible for a second and then seems to unmount but not cleanup correctly. The app is in a frozen state.

The fix for this was to just call hide action sheet in an async call back and await for it to close before hiding the modal:
```
  <DPButton
          testID='cancelSheetCancelButton'
          title={t('signup.cancelSheet.leave')}
          onPress={async () => {
            await hideActionSheet(SheetManagerSheets.cancel);
            displayConfirmUserInformationModal(false, dispatch);
          }}
          outline
        />
```

**Comment by ammarahm-ed** [comment]
A proper way is this:

```jsx
<Modal>
  <SheetProvider context="local" > 
   {children} // some child makes a call to open an action sheet
</SheetProvider>
</Modal>
```
In registering the sheet:

```js
registerSheet('mysheet',MySheet, 'global','local');
```
Then call open close with the contextId in payload:
```js
SheetManager.show("my-sheet",{ context:"local"})
SheetManager.hide("my-sheet",{ context:"local"})
```

**Comment by johnny-mcfadden-dailypay** [comment]
Ahh!! I seen the context params in the docs but hadn't fully looked into trying that yet. Brilliant!! Thank you @ammarahm-ed

**Comment by jordanmkoncz** [comment]
I ran into the same issue when trying to load a sheet from within a react-navigation screen with `presentation: "modal"`. I added an extra `SheetProvider` within that modal screen as the top-level element as per your suggestion @ammarahm-ed and it worked.

In order to make the sheet work within this screen and also within the regular global context from another (non-modal) screen, I had to pass through the dynamic `context` value as another variable in the custom payload to the sheet. It would be handy if `react-native-actions-sheet` provided some built-in way to access the `context` value used for the `SheetManager.show` call when in the sheet component, so that it doesn't have to be additionally passed manually in the payload.
